### ISSUE TITLE
Use image metadata to lay out images before the complete image data is available

### ISSUE THREAD

**Issue by jdm** [servo/servo#7047]
The [immeta library](https://github.com/netvl/immeta.git) can provide metadata about an image given a buffer. We should use this to retrieve the dimensions of an image as soon as possible while the image is downloading.

**Comment by jdm** [comment]
This is equivalent to Gecko's SIZE_AVAILABLE notification for imagelib.

**Comment by metajack** [comment]
Note that Seth has told me that this reading of image metadata (and dealing with embedded JPEG screenshots) is a prime target for Rust in Geckoness.

**Comment by jdm** [comment]
@jmr0 said they're interested in looking into this.

**Comment by jdm** [comment]
Relevant code to look at for servo involves ImageListener::notify in components/net/image_cache_task.rs, and ImageResponse in components/net_traits/image_cache_task.rs. We'll want a way to notify consumers (presumably via another entry in ImageResponse) that the metadata was successfully retrieved. The logic in HTMLImageElement::update_image will need to become more complex, and HTMLImageElement will need a separate field for the metadata that we care about. Finally, the ImageFragmentInfo code in components/layout/fragment.rs is where we'll end up using the new metadata.

**Comment by jmr0** [comment]
@jdm I take it the goal is for this metadata to be immediately accessible to the layout task? I was looking at this today through the debugger and noticed the image fragment's metadata is, as an example, accessed in `layout/inline.rs` through a call to `assign_replaced_inline_size`. A few steps after this there's an attempt to add a DisplayItem to a display list if the image data is available within the fragment. 

I'm curious as to whether modifying the image fragment's API (`image_block_size` and `image_inline_size`) to work with the pre-fetched metadata will right away yield some benefit, or if we should be looking at some other piece to avoid reflows?

Also, I noticed that ImageResponse is used in other places (like `dom/webglrenderingcontext.rs` and `dom/canvasrenderingcontext2d.rs`) and wasn't sure if this work would be applicable to those. Should a pre-fetched metadata response be ignored in such places (for now, at least)?

**Comment by jdm** [comment]
@jmr0 Yes, the code in `image_inline_size` and `image_block_size` should use the metadata if it's available, instead of deferring to the sizes present in the complete image data as in the current code.

You are also correct, the users in the webgl and canvas code are specifically concerned with the pixel data, so the metadata won't be enough for them.

**Comment by jmr0** [comment]
I think I've made some progress. At this point, HTMLImageElement and ImageFragment receive notifications that the metadata is available, and in the fragment's case it'll respond to `image_inline_size` and `image_block_size` using it. The biggest changes were to the image cache API, as I added support for just requesting an image or for separate metadata+image messages if metadata is available first -- this made it easier to handle any users that didn't care for the metadata alone.

I wasn't sure what this [block of code](https://github.com/jmr0/servo/commit/83be712354f00d1a18726a3cc64da2d718ac45f6#diff-1efd9021f44e3c1c1157af45f3ff140fR226) was about -- seemed to be for testing or sending requests in sync mode. I'm currently handling it by only requesting a full image and not its metadata if we land here.

I'm also wondering if this is the best way to handle [this](https://github.com/jmr0/servo/commit/83be712354f00d1a18726a3cc64da2d718ac45f6#diff-90a6391c1c7c20ffbfca3665f13cd657R351) (whether to perform the image metadata scan in another task as that would involve some synchronization around PendingLoad's byte buffer or cloning).

I'm not entirely sure how HTMLImageElement can use the metadata directly yet -- I'll take a look at the spec soon and probably send some more questions your way. Thanks again :)

The whole commit is here: https://github.com/jmr0/servo/commit/83be712354f00d1a18726a3cc64da2d718ac45f6
